modules/dataframe/src/io/csvreader.cpp

```diff
@@ -484,32 +495,106 @@ std::vector<std::function<void(std::string_view, size_t, size_t)>> CSVReader::ad
+                        }
+                        return neutralValue;
+                    },
+                    [&](const std::function<bool(double)>& func) {
+                        if (auto val = util::toNumber<double>(cell, cLocale)) {
+                            return func(*val);
+                        }
+                        return neutralValue;
+                    }};
+                for (const auto& f : filters) {
+                    if ((!firstRowHeader_ || (f.filterOnHeader == filterOnHeader)) &&
+                        f.column == static_cast<int>(colIndex)) {
+                        retval |= std::visit(test, f.filter);
+                    }
+                }
+            });
+        return retval;
+    };
+
+    auto filterApplicable = [&](auto& filters) {
+        return !filters.empty() &&
+               (!firstRowHeader_ || std::any_of(filters.begin(), filters.end(), [&](auto& f) {
+                   return (f.filterOnHeader == filterOnHeader);
+               }));
+    };
+
+    if (filterApplicable(filters_.includeRows) &&
+        std::none_of(filters_.includeRows.begin(), filters_.includeRows.end(), filterRow(true))) {
+        return true;
+    }
+    if (std::any_of(filters_.excludeRows.begin(), filters_.excludeRows.end(), filterRow(false))) {
+        return true;
+    }
+
+    if (filterApplicable(filters_.includeItems) && !filterItems(filters_.includeItems, true)) {
+        return true;
+    }
+    if (filterItems(filters_.excludeItems, false)) {
+        return true;
+    }
+
+    return false;
+}
+
 std::shared_ptr<DataFrame> CSVReader::readData(std::istream& stream) const {
     filesystem::skipByteOrderMark(stream);
 
     util::OnScopeExit cleanup{nullptr};
     if (!config::charconv || locale_ != "C") {
         // We need to use the C locale here to force use of decimal "."
-        std::string prev{std::setlocale(LC_ALL, nullptr)};
+        auto prevLocale = std::setlocale(LC_ALL, nullptr);
+        std::string prev{prevLocale ? prevLocale : ""};
         if (!std::setlocale(LC_ALL, locale_.c_str())) {
             LogWarn("Failed to set locale " << locale_);
         }
         cleanup.setAction([prev]() { std::setlocale(LC_ALL, prev.c_str()); });
     }
 
     std::string content{std::istreambuf_iterator<char>(stream), std::istreambuf_iterator<char>()};
+    std::string_view trimmed{content};
+    if (auto pos = trimmed.find_last_not_of(" \f\n\r\t\v"); pos != std::string_view::npos) {
+        trimmed = trimmed.substr(0, pos + 1);
+    }
 
     std::vector<std::pair<std::string_view, size_t>> rows;
-    util::parse(content, "\n", std::nullopt, std::nullopt,
+    util::parse(trimmed, "\n", std::nullopt, std::nullopt,
                 [&](std::string_view line, [[maybe_unused]] size_t index, size_t lineNumber) {
-                    rows.emplace_back(line, lineNumber);
+                    if (!skipRow(line, lineNumber, true)) {
+                        rows.emplace_back(line, lineNumber);
```

Comment:
Can't we do the second parsing here instead of below? Would avoid intermediate append and two skipRow calls.
Two skipRow calls should not be necessary?
```
util::parse(line, delimiters_, headers.size(), lineNumber,
                        [&, l = lineNumber](std::string_view cell, size_t index,
                                            [[maybe_unused]] size_t part) {
                            appenders[index](cell, l, index + 1);
                        });
```